peer/persistentpeer_mgr.go

```diff
@@ -0,0 +1,621 @@
+	// avoid the nodes entering connection cycles.
+	margin := nextBackoff / 10
+
+	var wiggle big.Int
+	wiggle.SetUint64(uint64(margin))
+	if _, err := rand.Int(rand.Reader, &wiggle); err != nil {
+		// Randomizing is not mission critical, so we'll just return the
+		// current backoff.
+		return nextBackoff
+	}
+
+	// Otherwise add in our wiggle, but subtract out half of the margin so
+	// that the backoff can tweaked by 1/20 in either direction.
+	return nextBackoff + (time.Duration(wiggle.Uint64()) - margin/2)
+}
+
+// ConnectPeer uses all the stored addresses for a peer to attempt to connect
+// to the peer. It creates connection requests if there are currently none for
+// a given address, and it removes old connection requests if the associated
+// address is no longer in the latest address list for the peer.
+func (m *PersistentPeerManager) ConnectPeer(pubKey *btcec.PublicKey) {
+	m.Lock()
+	defer m.Unlock()
+
+	peer, ok := m.conns[route.NewVertex(pubKey)]
+	if !ok {
+		peerLog.Debugf("Peer %x is not a persistent peer. Ignoring "+
+			"connection attempt", pubKey.SerializeCompressed())
+		return
+	}
+
+	if len(peer.addrs) == 0 {
+		peerLog.Debugf("Ignoring connection attempt to peer %s "+
+			"without any stored address",
+			pubKey.SerializeCompressed())
+		return
+	}
+
+	// Create an easy lookup map of the addresses we have stored for the
+	// peer. We will remove entries from this map if we have existing
+	// connection requests for the associated address and then any leftover
+	// entries will indicate which addresses we should create new
+	// connection requests for.
+	addrMap := make(map[string]*lnwire.NetAddress)
+	for key, addr := range peer.addrs {
+		addrMap[key] = addr
+	}
+
+	// Go through each of the existing connection requests and
+	// check if they correspond to the latest set of addresses. If there is
+	// a connection requests that does not use one of the latest advertised
+	// addresses then remove that connection request.
+	var updatedConnReqs []*connmgr.ConnReq
+	for _, connReq := range peer.connReqs {
+		lnAddr := connReq.Addr.(*lnwire.NetAddress).String()
+
+		switch _, ok := addrMap[lnAddr]; ok {
+		// If the existing connection request is using one of the
+		// latest advertised addresses for the peer then we add it to
+		// updatedConnReqs and remove the associated address from
+		// addrMap so that we don't recreate this connReq later on.
+		case true:
+			updatedConnReqs = append(
+				updatedConnReqs, connReq,
+			)
+			delete(addrMap, lnAddr)
+
+		// If the existing connection request is using an address that
+		// is not one of the latest advertised addresses for the peer
+		// then we remove the connecting request from the connection
+		// manager.
+		case false:
+			peerLog.Info(
+				"Removing conn req:", connReq.Addr.String(),
+			)
+			m.cfg.ConnMgr.Remove(connReq.ID())
+		}
+	}
+
+	peer.connReqs = updatedConnReqs
+	cancelChan := peer.getRetryCanceller()
```

Comment:
I think there may be a race condition with running goroutines created by the previous `ConnectPeer` that are blocked on the mutex?

Reply:
with `peer.connReqs`?